### ISSUE TITLE
session 4. NN learning 

### ISSUE THREAD

**Issue by Odin-son** [Odin-son/felis-catus#3]

**Comment by Odin-son** [comment]
end-to-end 방식은 '처음부터 끝까지' 데이터(입력)에서 목표로하는 결과(출력)까지 사람의 개입 없다는 것을 의미

**Comment by Odin-son** [comment]
신경망도 하나의 지표를 기준으로 최적의 매개변수 값을 탐색
이러한 지표를 손실함수, loss function, cost function 라고 하며,
일반적으로 SSE와 CEE를 사용

**Comment by Odin-son** [comment]
미니배치(mini-batch) - 훈련 데이터로부터 일부만 골라 학습을 수행
np.random.choice(60000,100) # 60000미만의 수 중 100개만 무작위로 추출

**Comment by Odin-son** [comment]
y[np.arange(batch_size),t] => y[0,2] ... (5,)의 shape이 나옴

**Comment by Odin-son** [comment]
손실함수를 왜 사용해야 하는가. => 정확도 높이기 위해서
그러면 정확도만 가지고 하면 되지 않는가...?
최적의 매개변수를 찾을때, 매개변수의 미분을 찾아 이를 근거로 값을 서서히 갱신하기 때문
정확도를 매개변수로 한다면, 대부분의 장소에서 0 이 되기때문에 더이상 갱신하지않음
계단함수와 시그모이드 함수를 생각해보면 좋을듯

**Comment by Odin-son** [comment]
f(x+h)-f(x) / h => 전방차분
(f(x+h)-f(x-h)) / (2*h) => 중심차분, 중앙차분 = x를 중심으로 그 전후의 차분을 계산한다는 의미

**Comment by Odin-son** [comment]
결국, 기울기는 각 지점에서 낮아지는 방향을 가리킴

**Comment by Odin-son** [comment]
하이퍼 파라미터는 학습률(eta)처럼 사람이 직접 설정해줘야하는 매개변수

**Comment by Odin-son** [comment]
SGD(stochastic gradient descent) : 확률적으로 무작위로 골라낸(mini-batch) 데이터 에 대해 수행하는 경사 하강법

**Comment by Odin-son** [comment]
* common.py 생성
  * sigmoid, softmax, identity_function, cross_entropy_error

**Comment by Odin-son** [comment]
경사하강법(gradient-descent) 보다 오차역전파법이 빠름

**Comment by Odin-son** [comment]
epoch , 10000개의 데이터에서 100개로 미니배치를 수행하였을때 100회 수행한게 1epoch

**Comment by Odin-son** [comment]
오버피팅을 피하기 위해 정확도가 떨어지는 순간 조기 종료(early stopping) -> 6.4에서 배움(가중치 감소, 드롭아웃)
